src/libraries/FluidLib.sol

```diff
@@ -0,0 +1,196 @@
+    /// @return amount_ The amount of tokens minted for the deposit
+    /// @return shares_ The number of dex pool shares deposited
+    function deposit(
+        uint256 token0Amt_,
+        uint256 token1Amt_,
+        uint256 minSharesAmt_,
+        address to_
+    ) external payable returns (uint256 amount_, uint256 shares_);
+
+    /// @dev This function allows users to withdraw tokens in any proportion from the col pool
+    /// @param token0Amt_ The amount of token0 to withdraw
+    /// @param token1Amt_ The amount of token1 to withdraw
+    /// @param maxSharesAmt_ The maximum number of shares the user is willing to burn
+    /// @param to_ Recipient of withdrawn tokens. If to_ == address(0) then out tokens will be sent to msg.sender. If to_ == ADDRESS_DEAD then function will revert with shares_
+    /// @return amount_ The number of tokens burned for the withdrawal
+    /// @return shares_ The number of dex pool shares withdrawn
+    function withdraw(
+        uint256 token0Amt_,
+        uint256 token1Amt_,
+        uint256 maxSharesAmt_,
+        address to_
+    ) external returns (uint256 amount_, uint256 shares_);
+
+    function TOKEN0() external returns (address);
+    function TOKEN1() external returns (address);
+}
+
+library FluidLib {
+    /**********************************************************************************************/
+    /*** Structs                                                                                ***/
+    /**********************************************************************************************/
+
+    struct DepositSmartLendingParams {
+        IALMProxy proxy;
+        IRateLimits rateLimits;
+        address smartLending;
+        uint256 token0Amount;
+        uint256 token1Amount;
+        uint256 minShares;
+        bytes32 depositSLRateLimitId;
+    }
+
+    struct WithdrawSmartLendingParams {
+        IALMProxy proxy;
+        IRateLimits rateLimits;
+        address smartLending;
+        uint256 token0Amount;
+        uint256 token1Amount;
+        uint256 maxShares;
+        bytes32 withdrawSLRateLimitId;
+    }
+
+    /**********************************************************************************************/
+    /*** External functions                                                                     ***/
+    /**********************************************************************************************/
+
+    function depositSmartLending(
+        DepositSmartLendingParams calldata params
+    ) external returns (uint256 shares) {
+        // only supports ERC20s. native ETH support can be added if needed.
+
+        IFluidSmartLending smartLending = IFluidSmartLending(
+            params.smartLending
+        );
+
+        if (params.token0Amount > 0) {
+            address token0 = smartLending.TOKEN0();
+
+            params.rateLimits.triggerRateLimitDecrease(
+                RateLimitHelpers.makeAssetDestinationKey(
```

Comment:
is this ok? this splits the rate limit uniquely into a SmartLending pool and then further into each token0 and token1